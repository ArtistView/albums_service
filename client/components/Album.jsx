import React from 'react';
import ReactDom from 'react-dom';

const Album = (props) => {
  console.log('hello world');
  return <div>Hello World</div>;
};

export default Album;