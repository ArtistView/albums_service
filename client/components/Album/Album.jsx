import React from 'react';
import $ from 'jquery';
import styled, { css } from 'styled-components';
import {GrCirclePlay} from 'react-icons/gr';

const AlbumCover = styled.img`
  width:173.25px;
  height:173.25px;
  object-fit: cover;
`;

const PlayButton = styled.img`
  display: none;
  position: absolute;
  right: 0;
  left: 0;
  top: 0;
  bottom: 0;
  margin: auto auto;
  width:70px;
  height:70px;
  &:hover {
    width:73px;
    height:73px;
  };
`; // change percentages and size as needed and make sure it's centered

const AlbumTitle = styled.div`
  font-size: small;
  margin-top: 10px;
  &:hover {
    text-decoration: underline;
  };
  display: inline-block;
`;

const AlbumCoverWrapper = styled.div`
  position: relative;
  width:173.25px;
  height:173.25px;
`;

// resize to fit the 232 offset margin
const AlbumWrapper = styled.div`
  position: relative;
  text-align: center;
  float: left;
  margin-left: 11px;
  margin-right: 11px;
  margin-bottom: 50px;
  width: 173.25px;
  &:hover ${AlbumCover} {
    opacity: .5;
  };
  &:hover ${PlayButton} {
    display: block;
  };
`;

class Album extends React.Component {
  // takes in an album object as a prop as well as a startPlaying function, which modifies the app state for which album is playing and a currPlaying property which stores the app state for which album is playing
  constructor(props) {
    super(props);
    this.myRef = React.createRef();
    this.state = {
      playing: false, // tracking if the album is currently playing
      buttonUrls: ['https://fakespotify.s3-us-west-1.amazonaws.com/play-button-2.png', 'https://fakespotify.s3-us-west-1.amazonaws.com/pause-button-try-3.png'], // array of the play and puase button images from my S3
      // buttonUrls: ['https://www.pngfind.com/pngs/m/427-4277341_add-play-button-to-image-online-overlay-play.png', 'https://tinyurl.com/ybdhlhsl'], // array of play and pause button images not through S3, doesn't look like spotify but it won't charge me, uncomment this and comment the line above for SDC bc those images won't be available
      // functionality remains intact
      // list of links to my play and pause icons
      buttonIndex: 0, // which icon (play (0) or pause(1)) is currently showing, used to access the links from buttonUrls array
      currSongIndex: 0, // stores the index of the current song to be played within the songs array of the album
      audio: new Audio(''), // stores the mp3 file for the song that is playing
      audioList: [], // stores a list of all xthe mp3 files as audios for songs in the album
    };
    this.play = this.play.bind(this);
    this.getSongs = this.getSongs.bind(this);
    this.playNext = this.playNext.bind(this);
    this.updateSizing = this.updateSizing.bind(this);
  }

  componentDidMount() {
    this.getSongs(this.props.album._id);
    // upon loading the component it gets the list of mp3 and sets them in state
    window.addEventListener('resize', this.updateSizing);
    // fills the albums to the window as the size changes
  }

  componentDidUpdate() {
    // this will run whenever a prop is changed (which will happen whenever the currPlaying changes to another album by play being clicked in that component)
    // makes sure that when another album is clicked, the current album stops playing and resets its behavior to what it is initially
    if(this.props.currPlaying !== this.props.album._id && this.props.currPlaying !== '' && this.state.playing === true) {
      this.refs.playbutton.style.display = ''; // makes the play button return to normal hover behavior as we set in css
      this.refs.albumcover.style.opacity = '';
      // reset the song state to the first song in the album
      this.state.audio.pause(); // pauses the audio upon another album's click
      this.setState({
        playing: false,
        buttonIndex: 0,
        audio: new Audio(this.state.audioList[0]),
      });
      this.state.audio.currentTime = 0; // resets the song to the beginning
      this.state.audio.addEventListener('ended', () => {
        this.playNext();
      });
      // need to test this stuff
    }
    if(this.props.currPlaying === this.props.album._id && this.props.show === true && this.state.playing === true) { // if this is playing but was collapsed through the show less button
      this.refs.albumcover.style.opacity = 0.5;
      this.refs.playbutton.style.display = 'block';
      // this changes it to showing but it keeps these style settings permanently, won't change it back to the defaults when I press pause
    }
    if(this.props.show === true) {
      this.updateSizing(); // changes the sizing to match the screen size
    }
  }

  getSongs(album) {
    const url = `http://localhost:3273/songs/byAlbum/${album}`;
    fetch(url) // fetches the list of mp3s for the given album
      .then((response) => response.json())
      .then((data) => {
        this.setState({
          audioList: data,
        });
      })
      .then(() => {
        this.setState({
          audio: new Audio(this.state.audioList[this.state.currSongIndex]),
        });
        this.state.audio.addEventListener('ended', () => {
          this.playNext(); // add event listener to the audio that plays the next song once it ends
        });
      })
      .catch((err) => {
        console.error(err);
      });
  }

  play(event) {
    // on click of the play button, we toggle the play and pause images and implement functionality and styling
    // also plays/pauses audio on click
    if (this.state.playing === false) {
      event.target.style.display = 'block'; // makes the pause button persist even after hover
      this.props.startPlaying(this.props.album._id); // changes the global state to be playing this album
      // this will trigger any other album that is playing to stop through the componentDidUpdate method
      this.refs.albumcover.style.opacity = 0.5; // the album cover will stay darker when the pause is showing as if it was being hovered over
      this.state.audio.play(); // plays the audio from the mp3 file for the current song
      this.setState({
        playing: true,
        buttonIndex: 1,
      });
    } else {
      event.target.style.display = '';
      this.refs.albumcover.style.opacity = '';
      // the above two lines make the play button return to normal hover behavior as we set in css
      this.state.audio.pause(); // pauses the current mp3 file upon clicking the pause button
      this.setState({
        playing: false,
        buttonIndex: 0,
      });
    }
  }

  playNext() {
    // this method will go to the next song in the album list and play it
    this.setState({ // increments the current song before checking if the album has ended
      currSongIndex: this.state.currSongIndex + 1,
    });
    if (this.state.currSongIndex >= this.state.audioList.length) {
      // if there are no more songs left in the album, reset the current song to the beginning of the album but don't play it
      this.setState({
        playing: false,
        buttonIndex: 0,
        audio: new Audio(this.state.audioList[0]),
        currSongIndex: 0,
      });
      this.refs.playbutton.style.display = '';
      this.refs.albumcover.style.opacity = '';
      this.state.audio.addEventListener('ended', () => {
        this.playNext();
      });
      // makes the play button return to normal hover behavior as we set in css
    } else {
      // if there are more songs in the list play the next song (which was incremented before the else)
      this.setState({
        audio: new Audio(this.state.audioList[this.state.currSongIndex]),
      });
      this.state.audio.play();
      this.state.audio.addEventListener('ended', () => {
        this.playNext();
      });
    }
  }

  updateSizing() {
    // set number to be the number of elements shown in the first two rows
    let percentPerAlbum = 0;
    const width = $(window).width();
    if (width > 1040 && width < 1200) {
      percentPerAlbum = 21;
    } else if (width < 1040 && width > 850) {
      percentPerAlbum = 28;
    } else if (width < 850) {
      percentPerAlbum = 43;
    } else {
      this.refs.albumwrapper.style.width = '';
      this.refs.albumcoverwrapper.style.height = '';
      this.refs.albumcoverwrapper.style.width = '';
      this.refs.albumcover.style.height = '';
      this.refs.albumcover.style.width = '';
      return;
    }
    let size = (width - 232) * (percentPerAlbum / 100);
    this.refs.albumwrapper.style.width = size;
    this.refs.albumcoverwrapper.style.height = size;
    this.refs.albumcoverwrapper.style.width = size;
    this.refs.albumcover.style.height = size;
    this.refs.albumcover.style.width = size;
  }

  render() {
    // renders an album cover with a play button on top of it that has clickable functionality
    // also renders an album title underneath that underlines on hover but doesn't have clickable functionality as that is outside the scope of the artist page, theoretically in actual spotify a click will take you to the album's page though
    if (this.props.show) { // if it should be showing
      return (
        <AlbumWrapper ref="albumwrapper" className="album" id={this.props.id}>
          <AlbumCoverWrapper ref="albumcoverwrapper" id="album-cover-wrapper">
            <AlbumCover ref="albumcover" id="album-cover" src={this.props.album.imageUrl} alt="" />
            <PlayButton ref="playbutton" id="play-button" src={this.state.buttonUrls[this.state.buttonIndex]} alt="" onClick={this.play} />
          </AlbumCoverWrapper>
          <AlbumTitle id="album-title">
            {this.props.album.title}
          </AlbumTitle>
        </AlbumWrapper>
      );
    }
    return (<div />);
  }
}

export default Album;
