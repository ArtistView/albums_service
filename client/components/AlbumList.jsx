import React from 'react';
import ReactDom from 'react-dom';
import Album from './Album.jsx';

const AlbumList = (props) => {
  console.log('hello world');
  return <div>Hello World</div>;
};

export default AlbumList;