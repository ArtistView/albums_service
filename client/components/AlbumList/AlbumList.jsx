import React from 'react';
import $ from 'jquery';
import Promise from 'bluebird';
import styled, { css } from 'styled-components';
import {AiOutlineDown, AiOutlineUp} from 'react-icons/ai';
import Album from '../Album/Album.jsx';

const AlbumListWrapper = styled.div`
  padding-left: 25px;
  background-color: #151515;
`;

const AlbumListTitle = styled.div`
  font-size: 28px;
  padding-top: 65px;
  margin-bottom: 20px;
  margin-left: 10px;
  margin-right: 10px;
`;

const AlbumListShowMoreLess = styled.div`
  letter-spacing: 2px;
  clear: both;
  text-align: center;
  margin-top: 20px;
  font-size: 11px;
  font-weight: normal;
  position: relative;
`;

class AlbumList extends React.Component {
  // this component accepts a list of albums as props along with a playing function and currPlaying variable that are just needed to pass down to the album component, so it can modify the global state of which album is playing
  // if the type is Appears on, the albums prop is already all the featured albums, otherwise it is all albums that are primarily by that artist regardless of type (thus we will filter by type in componentDidMount)
  constructor(props) {
    super(props);
    this.state = {
      type: this.props.type, // stores the type of album this list is showing
      albums: [], // stores the list of albums that match the type
      showMore: false, // boolean telling us whether or not we should show more than 12 albums
      showMoreText: '', // stores the text to be rendered within the show more/less button
      showMoreArrow: '', // stores the icon to be rendered with the show more/less button
      showAtAll: false, // boolean telling us whether to show the component at all, should be false only if there are no albums of this particular type
      currShowing: [], // stores the list of albums to show, could be all the albums or just the first 12 depending on showMore
      numToShow: 12, // stores the number of albums that should be shown when show more is not pressed, should change when the window shrinks
      // TODO: update numToShow when the window shrinks and resize the albums to fit the window
    };
    this.showMoreLess = this.showMoreLess.bind(this);
    this.updateNumShowing = this.updateNumShowing.bind(this);
    this.populateAlbums = this.populateAlbums.bind(this);
  }

  async componentDidMount() {
    // upon loading we loop through all the albums from the artist and add only the ones that match our type to the array, which is then set to be albums in state
    if ($(window).width() < 1200) {
      await this.updateNumShowing();
      // if the screen is smaller, update the number and wait for it to finish before moving on
    }
    this.populateAlbums();
    // populate the state with the correct albums
  }
  // need to write tests

  populateAlbums() {
    // populates the state with albums that match the type
    let albumsOfType = [];
    if (this.state.type === 'Appears On') {
      albumsOfType = this.props.albums;
    } else if (this.state.type === 'Albums') {
      for (var album of this.props.albums) {
        if (album.type === 'album') {
          albumsOfType.push(album);
        }
      }
    } else if (this.state.type === 'Singles and EPs') {
      for (var album of this.props.albums) {
        if (album.type === 'single' || album.type === 'EP') {
          albumsOfType.push(album);
        }
      }
    } else {
      for (var album of this.props.albums) {
        if (album.type === 'compilation') {
          albumsOfType.push(album);
        }
      }
    }
    // if there are any albums of this type, show them and set the list equal to albums
    if (albumsOfType.length > 0) {
      this.setState({
        albums: albumsOfType,
        showAtAll: true,
        currShowing: albumsOfType,
      });
    }
    // if there are more than 12, make sure the show more button appears and we start by only showing the first 12 albums
    if (albumsOfType.length > this.state.numToShow) {
      this.setState({
        showMoreText: 'SHOW MORE',
        currShowing: albumsOfType.slice(0, this.state.numToShow),
        showMoreArrow: <AiOutlineDown id="album-down-arrow" />,
      });
    }
    window.addEventListener('resize', this.updateNumShowing);
    // adds an event listener that updates the number of albums showing if the window resizes
  }

  updateNumShowing() {
    // set number to be the number of elements shown in the first two rows
    let number = 12;
    const width = $(window).width();
    if (width > 1040 && width < 1200) {
      number = 8;
    } else if (width < 1040 && width > 850) {
      number = 3;
    } else if (width < 850) {
      number = 2;
    } else {
      number = 12;
    }
    if (this.state.showMore === true) {
      this.setState({
        numToShow: number,
        currShowing: this.state.albums,
      });
    } else {
      this.setState({
        numToShow: number,
        currShowing: this.state.albums.slice(0, number),
      });
    }

  }

  showMoreLess(event) {
    if (this.state.showMore === false) {
      // if it's currently showing twelve (or less if the screen is smaller) make it show all of the albums and change the text
      this.setState({
        showMore: true,
        showMoreText: 'SHOW LESS',
        currShowing: this.state.albums,
        showMoreArrow: <AiOutlineUp id="album-up-arrow" />,
      });
    } else {
      // if it's currently showing all the albums make it only show 12 (numToShow) and change the text
      this.setState({
        showMore: false,
        showMoreText: 'SHOW MORE',
        currShowing: this.state.albums.slice(0, this.state.numToShow),
        showMoreArrow: <AiOutlineDown id="album-down-arrow" />,
      });
    }
  }

  render() {
    // the albums list component renders a title, which is the type, a list of album components and a show more/less button with clickable functionality
    if (this.state.showAtAll) {
      return(
        <AlbumListWrapper className="album-list">
          <AlbumListTitle className="album-list-title">{this.state.type}</AlbumListTitle>
          <div className="album-list-list">
            {this.state.albums.map((album) => {
              // want to rerun this map function when currShowing changes
              let showing = false;
              if (this.state.currShowing.includes(album)) {
                showing = true;
              }
              return (<Album album={album} key={album._id} id={album._id} startPlaying={this.props.playing} currPlaying={this.props.currPlaying} show={showing} />);
            })
            }
          </div>
          <AlbumListShowMoreLess className="album-list-show-more-less" onClick={this.showMoreLess}>
            {this.state.showMoreText}
            {this.state.showMoreArrow}
          </AlbumListShowMoreLess>
        </AlbumListWrapper>
      );
    }
    return <div id="no-albums-of-this-type" />;
  };
};

export default AlbumList;
