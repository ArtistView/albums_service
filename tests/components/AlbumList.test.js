/**
 * @jest-environment jsdom
*/
import React from 'react';
import sinon from 'sinon';
import ReactDom from 'react-dom';
import { shallow, mount, render, configure } from 'enzyme';
import AlbumList from '../../client/components/AlbumList/AlbumList.jsx';
import Album from '../../client/components/Album/Album.jsx';

const albums = [
  {
    artistId: '5eebe7b2c451ef0e3f4c0533',
    featuredArtists: ['5eebdbf1bf2d490c13ff868d'],
    imageUrl: 'https://tinyurl.com/yc9dd4jq',
    songs: ['5eebfe07a386ca12ede16e2f', '5eebfe07a386ca12ede16e93', '5eebfe07a386ca12ede16e2f', '5eebfe07a386ca12ede16e93', '5eebfe07a386ca12ede16e2f', '5eebfe07a386ca12ede16e93'],
    title: 'aliquam',
    type: 'album',
    __v: 0,
    _id: '5eebecb1969c0a0fa4e0ce45',
  },
];
const noop = () => {};
describe('AlbumList', () => {
  const wrapper = shallow(<AlbumList type="Albums" albums={albums} playing={albums[0]._id} currPlaying={noop} />);
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<AlbumList debug />);
    expect(component).toMatchSnapshot();
  });
  it('should render correctly with props', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('has state', () => {
    expect(wrapper.state('type')).toBe('Albums');
    expect(wrapper.state('showMore')).toBe(false);
    expect(wrapper.state('showMoreText')).toBe('');
    expect(wrapper.state('showMoreArrow')).toBe('');
  });
  it('recieves the correct props', () => {
    expect(wrapper.instance().props.albums).toEqual(albums);
    expect(wrapper.instance().props.playing).toBe(albums[0]._id);
    expect(wrapper.instance().props.currPlaying).toBe(noop);
    expect(wrapper.instance().props.type).toBe("Albums");
  });
  // the coverage is ok on this one but I pretty much cover all the rest of the lines and functions in my end to end testing through cypress
});
