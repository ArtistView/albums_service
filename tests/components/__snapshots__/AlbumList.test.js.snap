// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumList should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`AlbumList should render correctly with props 1`] = `ShallowWrapper {}`;
